msi/release.sh

```diff
@@ -1,34 +1,119 @@
 #!/bin/sh
 
-# Build Git for Windows' .msi files
+# Build Git for Windows .msi file
 
 test -z "$1" && {
-	echo "Usage: $0 <version> [optional components]"
-	exit 1
+    echo
+    echo "Usage: $0 [-v=|--version=|-v ]<version> [(-o=|--outputDir=|-o ) <Output Directory>] [Optional Components]"
+    echo
+    echo "   -v <version>"
+    echo "   -v=<version>"
+    echo "   --version=<version>"
+    echo "       The version of Git for Windows for which to create an installer."
+    echo "       The version should be in the format:"
+    echo
+    echo "           <Major>.<Minor>.<Revision>.<Build/Patch>."
+    echo
+    echo "   -o <Output Directory>"
+    echo "   -o=<Output Directory>"
+    echo "   --outputDir=<OutputDirectory>"
+    echo "       The location the completed installer should be written to."
+    echo "       If this option is not specified, the installer will be written to your"
+    echo "       home directory."
+    echo
+    echo "   [Optional Components]"
+    echo "       A list of additional packages you want bundled with your Git for Windows"
+    echo "       distribution installer."
+    echo
+    echo "       For example, if you want to install emacs with your Git for Windows"
+    echo "       installation, you would run the command:"
+    echo
+    echo "           ./release.sh M.m.r.b mingw-w64-emacs-git"
+    echo
+    echo "       where M.m.r.b is the Major, minor, revision, and build number, "
+    echo "       respectively."
+    echo
+    exit 1
 }
 
 die () {
-	echo "$*" >&1
-	exit 1
+    echo "$*" >&2
+    exit 1
+}
+
+function parse_version() {
+    echo "$1" | grep -Px '\d+(\.(0|[1-9]+)){2,3}'
 }
 
 ARCH="$(uname -m)"
 case "$ARCH" in
 i686)
-	BITNESS=32
-	WIX_ARCH=x86
-	;;
+    BITNESS=32
+    WIX_ARCH=x86
+    ;;
 x86_64)
-	BITNESS=64
-	WIX_ARCH=x64
-	;;
+    BITNESS=64
+    WIX_ARCH=x64
+    ;;
 *)
-	die "Unhandled architecture: $ARCH"
-	;;
+    die "Unhandled architecture: $ARCH"
+    ;;
 esac
-VERSION=$1
-shift
-TARGET="$HOME"/Git-"$VERSION"-"$BITNESS"-bit.msi
+
+for arg
+do
+    case "$1" in
+        -o=*|--outputDir=*)
+            TARGET="${1##*=}"
+            shift
+        ;;
+        -o)
+            shift
+            TARGET="$1"
+            shift
+        ;;
+        -v=*|--version=*)
+            VERSION="$(parse_version $(echo "${1##*=}"))"
+            shift
+        ;;
+        -v)
+            shift
+            VERSION="$(parse_version "$1")"
+            shift
```

Comment:
Since the version is *required*, it would be nicer to keep it as the first non-option, i.e.

```
Usage: msi/release.sh [-o <dir>] <version> [<optional-package>...]
```